lightning/backend/local.go

```diff
@@ -1609,186 +1730,307 @@ func (s *sizeProperties) iter(f func(p *rangeProperty) bool) {
+	sw.totalCount += int64(len(m.kvs))
+	sw.lastKey = m.kvs[len(m.kvs)-1].Key
+	m.clear()
+	return nil
+}
+
+// ingestInto finishes the SST file, and ingests itself into the target LocalFile database.
+// On success, the entire writer will be reset as empty.
+func (sw *sstWriter) ingestInto(e *LocalFile, desc localIngestDescription) error {
+	if sw.totalCount > 0 {
+		if err := sw.writer.Close(); err != nil {
+			return errors.Trace(err)
+		}
+		if desc&localIngestDescriptionFlushed == 0 {
+			// No need to acquire lock around ingestion when flushing.
+			// we already held the lock before flushing.
+			e.lock(importMutexStateLocalIngest)
+			defer e.unlock()
+		}
+		meta, _ := sw.writer.Metadata() // this method returns error only if it has not been closed yet.
+		log.L().Info("write data to local DB",
+			zap.Int64("size", sw.totalSize),
+			zap.Int64("kvs", sw.totalCount),
+			zap.Uint8("description", uint8(desc)),
+			zap.Uint64("sstFileSize", meta.Size),
+			log.ZapRedactBinary("firstKey", meta.SmallestPoint.UserKey),
+			log.ZapRedactBinary("lastKey", meta.LargestPoint.UserKey))
+
+		if err := e.db.Ingest([]string{sw.path}); err != nil {
+			return errors.Trace(err)
+		}
+		e.TotalSize.Add(sw.totalSize)
+		e.Length.Add(sw.totalCount)
+		sw.totalSize = 0
+		sw.totalCount = 0
+		sw.lastKey = nil
+	}
+	sw.writer = nil
+	return nil
+}
+
+// reopen creates a new SST file after ingestInto is successful.
+// Returns error if the SST file was not ingested.
+func (sw *sstWriter) reopen() error {
+	if sw.writer != nil {
+		return errors.New("cannot reopen an SST writer without ingesting it first")
+	}
+	f, err := os.Create(sw.path)
 	if err != nil {
-		return nil, filePath, err
+		return errors.Trace(err)
 	}
-	writer := sstable.NewWriter(f, sstable.WriterOptions{
+	sw.writer = sstable.NewWriter(f, sstable.WriterOptions{
 		TablePropertyCollectors: []func() pebble.TablePropertyCollector{
-			func() pebble.TablePropertyCollector {
-				return newRangePropertiesCollector()
-			},
+			newRangePropertiesCollector,
 		},
 		BlockSize: 16 * 1024,
 	})
-	return writer, filePath, nil
+	return nil
 }
 
-func writeKVs(writer *sstable.Writer, kvs []common.KvPair) error {
-	internalKey := sstable.InternalKey{
-		UserKey: []byte{},
-		Trailer: uint64(sstable.InternalKeyKindSet),
+// cleanUp removes any un-ingested SST file.
+func (sw *sstWriter) cleanUp() {
+	if sw.writer != nil {
+		sw.writer.Close()
+		os.Remove(sw.path)
 	}
-	for _, p := range kvs {
-		internalKey.UserKey = p.Key
-		if err := writer.Add(internalKey, p.Val); err != nil {
-			return err
+}
+
+// kvMemCache is an array of KV pairs. It also keep tracks of the total KV size and whether the array is already sorted.
+type kvMemCache struct {
+	kvs       []common.KvPair
+	totalSize int64
+	notSorted bool // record "not sorted" instead of "sorted" so that the zero value is correct.
+}
+
+// append more KV pairs to the kvMemCache.
+func (m *kvMemCache) append(kvs []common.KvPair) {
+	if !m.notSorted {
```

Comment:
Even if the kvMemCache has been sorted, we shall still check the order of `kvs`. The array `kvs` may be unordered and the first key of `kvs` may be greater than the lastkey of `m.kvs`

Reply:
the loop inside does check the order of `kvs`.

```go
		var lastKey []byte
		if len(m.kvs) > 0 {
			lastKey = m.kvs[len(m.kvs)-1].Key
		}
		for _, kv := range kvs {
			if bytes.Compare(kv.Key, lastKey) <= 0 {    // <----
				m.notSorted = true
				break
			}
			lastKey = kv.Key   // <----
		}
```